packages/flutter/test/widgets/selectable_text_test.dart

```diff
@@ -0,0 +1,3634 @@
+            SemanticsAction.moveCursorForwardByCharacter,
+            SemanticsAction.moveCursorBackwardByWord,
+            SemanticsAction.moveCursorForwardByWord,
+            SemanticsAction.setSelection,
+          ],
+          flags: <SemanticsFlag>[
+            SemanticsFlag.isReadOnly,
+            SemanticsFlag.isTextField,
+            SemanticsFlag.isFocused,
+          ],
+        ),
+      ],
+    ), ignoreTransform: true, ignoreRect: true));
+
+    controller.selection = const TextSelection.collapsed(offset: 0);
+    await tester.pump();
+
+    expect(semantics, hasSemantics(TestSemantics.root(
+      children: <TestSemantics>[
+        TestSemantics.rootChild(
+          id: 1,
+          textDirection: TextDirection.ltr,
+          textSelection: const TextSelection.collapsed(offset: 0),
+          value: 'Guten Tag',
+          actions: <SemanticsAction>[
+            SemanticsAction.tap,
+            SemanticsAction.longPress,
+            SemanticsAction.moveCursorForwardByCharacter,
+            SemanticsAction.moveCursorForwardByWord,
+            SemanticsAction.setSelection,
+          ],
+          flags: <SemanticsFlag>[
+            SemanticsFlag.isReadOnly,
+            SemanticsFlag.isTextField,
+            SemanticsFlag.isFocused,
+          ],
+        ),
+      ],
+    ), ignoreTransform: true, ignoreRect: true));
+
+    semantics.dispose();
+  });
+
+  testWidgets('SelectableText semantics, enableInteractiveSelection = false', (WidgetTester tester) async {
+    final SemanticsTester semantics = SemanticsTester(tester);
+    final Key key = UniqueKey();
+
+    await tester.pumpWidget(
+      overlay(
+        child: SelectableText(
+          'Guten Tag',
+          key: key,
+          enableInteractiveSelection: false,
+        ),
+      ),
+    );
+
+    await tester.tap(find.byKey(key));
+    await tester.pump();
+
+    expect(semantics, hasSemantics(TestSemantics.root(
+      children: <TestSemantics>[
+        TestSemantics.rootChild(
+          id: 1,
+          value: 'Guten Tag',
+          textDirection: TextDirection.ltr,
+          actions: <SemanticsAction>[
+            SemanticsAction.tap,
+            SemanticsAction.longPress,
+            // Absent the following because enableInteractiveSelection: false
+            // SemanticsAction.moveCursorBackwardByCharacter,
+            // SemanticsAction.moveCursorBackwardByWord,
+            // SemanticsAction.setSelection,
+            // SemanticsAction.paste,
+          ],
+          flags: <SemanticsFlag>[
+            SemanticsFlag.isReadOnly,
+            SemanticsFlag.isTextField,
+            // SelectableText act like a text widget when enableInteractiveSelection
+            // is false. It will not respond to any pointer event.
+            // SemanticsFlag.isFocused,
+          ],
+        ),
+      ],
+    ), ignoreTransform: true, ignoreRect: true));
+
+    semantics.dispose();
+  });
+
+  testWidgets('SelectableText semanti
```

Comment:
It seems like most of these tests were copied from material/text_field_test.dart.  Instead, would it be possible to reuse the code by passing something like a `useSelectableText` flag?  This has the added benefit of making it obvious that TextField and SelectableText need to have the exact same selection behavior.  In the future when someone needs to add or edit a test in TextField, or change its behavior, they will be more likely to realize that they need to do the same for SelectableText, and vice versa.  If you had to heavily modify many of these tests then maybe it's not possible, but if not it could be a big improvement.

Another thing I want to ask about is test coverage for selection on iOS.  Is there thorough coverage for the iOS selection behavior and appearance for SelectableText?  Is there any way we can also work to guarantee that it's in sync with TextField and CupertinoTextField selection behavior?

Reply:
yes there are a good portion of ios coverage in this file. As we discussed offline, I have to rewrite a lot of  the tests since there are layout/semantics/selection difference between selectabletext and text field, it will be hard to combine the test. In the future i plan to factor out selection, I should be able to refactor and combine the test as well.